lib/std/zon/stringify.zig

```diff
@@ -0,0 +1,1920 @@
+            return;
+        } else switch (val) {
+            inline else => |payload| {
+                return checkValueDepth(payload, child_depth);
+            },
+        },
+        .Optional => if (val) |inner| try checkValueDepth(inner, child_depth),
+        else => {},
+    }
+}
+
+fn expectValueDepthEquals(expected: usize, value: anytype) !void {
+    try checkValueDepth(value, expected);
+    try std.testing.expectError(error.MaxDepth, checkValueDepth(value, expected - 1));
+}
+
+test "std.zon checkValueDepth" {
+    try expectValueDepthEquals(1, 10);
+    try expectValueDepthEquals(2, .{ .x = 1, .y = 2 });
+    try expectValueDepthEquals(2, .{ 1, 2 });
+    try expectValueDepthEquals(3, .{ 1, .{ 2, 3 } });
+    try expectValueDepthEquals(3, .{ .{ 1, 2 }, 3 });
+    try expectValueDepthEquals(3, .{ .x = 0, .y = 1, .z = .{ .x = 3 } });
+    try expectValueDepthEquals(3, .{ .x = 0, .y = .{ .x = 1 }, .z = 2 });
+    try expectValueDepthEquals(3, .{ .x = .{ .x = 0 }, .y = 1, .z = 2 });
+    try expectValueDepthEquals(2, @as(?u32, 1));
+    try expectValueDepthEquals(1, @as(?u32, null));
+    try expectValueDepthEquals(1, null);
+    try expectValueDepthEquals(2, &1);
+    try expectValueDepthEquals(3, &@as(?u32, 1));
+
+    const Union = union(enum) {
+        x: u32,
+        y: struct { x: u32 },
+    };
+    try expectValueDepthEquals(2, Union{ .x = 1 });
+    try expectValueDepthEquals(3, Union{ .y = .{ .x = 1 } });
+
+    const Recurse = struct { r: ?*const @This() };
+    try expectValueDepthEquals(2, Recurse{ .r = null });
+    try expectValueDepthEquals(5, Recurse{ .r = &Recurse{ .r = null } });
+    try expectValueDepthEquals(8, Recurse{ .r = &Recurse{ .r = &Recurse{ .r = null } } });
+
+    try expectValueDepthEquals(2, @as([]const u8, &.{ 1, 2, 3 }));
+    try expectValueDepthEquals(3, @as([]const []const u8, &.{&.{ 1, 2, 3 }}));
+}
+
+/// Lower level control over stringification, you can create a new instance with `stringifier`.
+///
+/// Useful when you want control over which fields/items are stringified, how they're represented,
+/// or want to write a ZON object that does not exist in memory.
+///
+/// You can serialize values with `value`. To serialize recursive types, the following are provided:
+/// * `valueMaxDepth`
+/// * `valueArbitraryDepth`
+///
+/// You can also serialize values using specific notations:
+/// * `int`
+/// * `float`
+/// * `utf8Codepoint`
+/// * `slice`
+/// * `sliceMaxDepth`
+/// * `sliceArbitraryDepth`
+/// * `string`
+/// * `multilineString`
+///
+/// For manual serialization of containers, see:
+/// * `startStruct`
+/// * `startTuple`
+/// * `startSlice`
+///
+/// # Example
+/// ```zig
+/// var serializer = stringifier(writer, .{});
+/// var vec2 = try serializer.startStruct(.{});
+/// try vec2.field("x", 1.5, .{});
+/// try vec2.fieldPrefix();
+/// try serializer.value(2.5);
+/// try vec2.finish();
+/// ```
+pub fn Stringifier(comptime Writer: type) type {
```

Comment:
If you use `std.io.AnyWriter` you can avoid this being a generic type. I think that's probably the right move, since it will result in less bloat, and there's already a more optimal, streaming output, that also happens to have a simpler implementation. The json equivalent is `std.json.WriteStream`. In fact that's such a good abstraction I bet this stringification code could be written cleanly and optimally on top of (the zon equivalent of) it.